framework/src/araui/core/ui_core.c

```diff
@@ -0,0 +1,835 @@
+		fps_dt -= ms_per_frame;
+#endif
+
+		view = ui_view_get_current();
+		if (view) {
+			prev_view = view;
+			root = view->root;
+
+			if (_ui_process_widget_recur(root, dt) != UI_OK) {
+				UI_LOGE("error: processing widget recursively failed!\n");
+			}
+
+			ui_render_widget(root);
+		}
+
+		if (_ui_core_is_visible_global_widget()) {
+			_ui_process_widget_recur(g_global_widget_info[g_core.visible_event_type], dt);
+		}
+
+		_ui_redraw(dt);
+
+#if defined(CONFIG_UI_ENABLE_TOUCH)
+		_ui_core_dispatch_touch_event();
+#endif
+
+		ui_process_all_requests();
+	}
+
+	g_core.state = UI_CORE_STATE_STOP;
+
+	UI_LOGD("[%s] has been finished.\n", UI_CORE_THREAD_NAME);
+
+	return NULL;
+}
+
+bool ui_is_running(void)
+{
+	return (g_core.state != UI_CORE_STATE_STOP);
+}
+
+#if defined(CONFIG_UI_ENABLE_TOUCH)
+
+static void _ui_deliver_touch_event(ui_widget_body_t *widget, ui_touch_event_t touch_event, ui_coord_t coord)
+{
+	ui_widget_body_t *parent = UI_NULL;
+
+	// If there is locked target, send all event to the target only
+	// else, send event to the focus widget and its hookers.
+
+	if (g_core.locked_target) {
+		g_core.locked_target->touch_cb(g_core.locked_target, touch_event, coord);
+	} else {
+		widget->touch_cb(widget, touch_event, coord);
+
+		// send touch event to the all hookers also
+		parent = widget->parent;
+		while (parent) {
+			if (parent->is_hooker && parent->touch_cb) {
+				parent->touch_cb(parent, touch_event, coord);
+			}
+			parent = parent->parent;
+		}
+	}
+}
+
+/**
+ * @brief Called when some widget is locked as an event target.
+ * But, the locked target doesn't get the event.
+ */
+static void _ui_cancel_touch_event_except_locked_target(ui_widget_body_t *widget)
+{
+	ui_coord_t null_coord = {0, };
+	ui_widget_body_t *parent;
+
+	// The focus widget should receive the cancel touch event
+	// whether the focus widget is a hooker or not.
+	if (widget->touch_cb && widget != g_core.locked_target) {
+		widget->touch_cb(widget, UI_TOUCH_EVENT_CANCEL, null_coord);
+	}
+
+	// Only hooker widgets inside the focus widget's parent list,
+	// should receive the touch cancel event.
+	parent = widget->parent;
+	while (parent) {
+		if (parent->is_hooker && parent->touch_cb && parent != g_core.locked_target) {
+			parent->touch_cb(parent, UI_TOUCH_EVENT_CANCEL, null_coord);
+		}
+		parent = parent->parent;
+	}
+}
+
+static void _ui_core_handle_touch_event(ui_touch_event_t touch_event, ui_coord_t coord)
+{
+	ui_view_body_t *view;
+	ui_widget_body_t *root;
+	ui_global_widget_body_t *global;
+
+	view = ui_view_get_current();
+	if (!view) {
+		UI_LOGE("error: active view is null!\n");
+		return;
+	}
+
+	root = view->root;
+
+	switch (touch_event) {
+		case UI_TOUCH_EVENT_DOWN: {
+			// If touch down, reset all infomations about focus and touch
+			ui_core_unlock_touch_event_target();
+
+			// If there is the visible global widget,
+			// a touch event will be delivered to only that global widget.
+			if (_ui_core_is_visible_global_widget()) {
```

Comment:
When touch down comes but there's no reason to send event to view, you should consume current event to prevent get another event(move, up) from dispatcher.

Reply:
This is AraUI Framework's internal code. Not related to the input dispatch driver.

Reply:
I mean ui should request consume to dispatcher, to prevent wrong & useless touch up event.(down - up is pair based on id)

Reply:
Actually, that is related to the architecture policy.
For me, controlling the touch event is UI's role, even if the module is separated from the AraUI.
because touch is one of user interaction.